### ISSUE TITLE
perf(vulkan): retain FP16 KV without F32 expansion

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12388]
Parent campaign: #11572. Frozen comparison ladder: #12093. Physical execution gate: #12098.

Order: **5 of 20** — goal-critical spine, ordered by expected leverage and prerequisite structure.
Study receipt: `study_fea6c1bc31f94fcee01ec566289e7ca509c28e7a982068dbde17b36e33b333e3`. Evidence observed: 09/08/2026 13:45:00.

## For / Problem / Today / Better because / Witness
- **For:** fak-native Qwen3.8 on this machine's Ryzen AI Max+ 395 / Radeon 8060S (gfx1151, 64 GiB unified memory).
- **Problem:** The resident Vulkan KV contract expands cache state to F32, doubling UMA traffic before attention and competing with the Q4_K_M weight stream.
- **Today:** #12093 remains `NO_COMPARABLE_LOCAL_RESULT`; the sealed candidate/reference path is #12364 → #12365 → #12366 → #12367 → #12368 → #12369.
- **Better because:** Retain K and V as FP16 across one append-read cycle, with named F32 accumulators only where the quality oracle requires them and no F32 shadow cache.
- **Witness:** A fixed append/read fixture matches the F32 oracle within the declared bound, resident KV bytes fall by at least 45%, fallback_count=0, and physical promotion requires an exact-artifact same-envelope full-model A/B.

## Comparison and proof contract
The contextual dense Qwen3.8-27B Q4_K_M bars are 16.65 / 24.13 / 29.42 accepted output tok/s at c1/c4/c8; the 30.56–36.04 q38rocm rows remain challenge/watch evidence until matched. These numbers are context only. This issue cannot independently claim that fak exceeds another engine.

Physical promotion must use the same-commit sealed candidate/reference path above, exact model artifact and tensor inventory, prompt/output lengths, accepted-token denominator, driver/build/power identity, at least five interleaved repetitions, quality non-inferiority, and `fallback_count=0`. A microbenchmark can complete the software leaf but cannot unlock a product or comparison win.

## Field-borrow evidence
- **Pinned source:** https://github.com/lhl/strix-halo-testing/blob/5ee14b5b0ecfbd16b7fd0aa41a91581c32b33b61/llm-bench/README.md#L10-L18
- **What transfers:** The pinned Strix study reports roughly 215 GB/s measured bandwidth, making avoidable live-cache bytes a first-order cost; fak must prove its own byte and quality deltas.
- **Repo self-query:** `fak capabilities --json 'Vulkan FP16 KV without F32 expansion'` → ABSENT: self-query returned no on-point capability; raw inspection found F32 resident KV storage.
- **Exact fak seam:** `internal/compute/vulkan_kv.go:32-45,70-117; internal/compute/shaders/attention.comp:15-18`.

## Dependencies and scope fences
- **Coordinates with:** #12098, #12264, #12387 (R04).
- **Promotion requires:** #12098, #12264, #12387 (R04).
- **Pickup rule:** Open coordination and promotion edges do not block a bounded software or test slice. Use `Start blocked by: #N` only with source-bound proof that no executable disjoint slice remains.
- Native execution stays fak-owned. No silent llama.cpp or CPU fallback; llama.cpp is comparator-only.
- Preserve the exact Qwen3.8 artifact, prompt/output contract, accepted-token accounting, and quality gate.
- Scope is limited to: `internal/compute/vulkan_kv.go, internal/compute/vulkan_kv_test.go, internal/compute/shaders/attention.comp`. Adjacent scheduler, allocator, codec, or model-format work is excluded unless named above.

## Done condition
- Retain K and V as FP16 across one append-read cycle, with named F32 accumulators only where the quality oracle requires them and no F32 shadow cache.

## Deterministic witness
`go test -tags vulkan ./internal/compute -run TestVulkanFP16KVNoF32Expansion -count=1`

## Physical promotion gate
- A fixed append/read fixture matches the F32 oracle within the declared bound, resident KV bytes fall by at least 45%, fallback_count=0, and physical promotion requires an exact-artifact same-envelope full-model A/B.
- Retain or default the candidate only when its source-bound physical effect clears the run's declared noise bound (normally ≥5%); otherwise record the elimination and advance to the next ranked lever.

```routing
lane: strix-vulkan-fp16-kv
max_files: 3
max_steps: 6
class: goal-critical spine
```

**Comment by anthony-chaudhary** [comment]
Status: `BLOCKED` — dependency and scope contracts prevent an honest R05 implementation on current trunk.

Evidence (observed 2026-09-08):
- `DEPENDENCY_OPEN_12098`: #12098 remains open with `NOT_COMPARABLE`; its landed docs receipt earns zero physical comparison credit.
- `DEPENDENCY_OPEN_12264`: #12264 remains open; only routing/docs are present, with no `ValueView` implementation or witness on `origin/main`.
- `DEPENDENCY_OPEN_12387`: immediate predecessor R04 remains open; `origin/main` has no #12387 resolving commit and no `TestVulkanQwen38BlockFP16ActivationContract` witness.
- `SCOPE_CONTRACT_MISMATCH`: the named three-file lane cannot satisfy the literal append/read done condition on current trunk. `internal/compute/vulkan_ops.go` directly dequantizes/reads the `[]float32` scratch buffers, while the actual resident append cache is in `internal/compute/vulkan_kv_vulkan.go` (`growAppend` uses four-byte F32 elements; K/V views are F32). Both are outside this issue's scope/max-files fence. `attention.comp` also binds K/V as F32 SSBOs.
- Baseline deterministic selector: `CLAIM_TEST_UNRUN`. Native `go test -tags vulkan ./internal/compute -run TestVulkanFP16KVNoF32Expansion -count=1 -v` failed before selection because `libfakvulkan` was unavailable; the bounded WSL run produced no terminal receipt. No pass is claimed.
- Prior-art route: `fak sota amd-vulkan-sequence-attention --json` resolves to borrow-with-oracle discipline (exact token identity, activation/logit parity, fak-native Vulkan identity, zero fallback). No hardware or speedup claim is made.

No repository files were edited or committed. Resume after all three dependencies land, re-read R04's resulting FP16 ABI, then either (a) amend this scope to include the real append/read seams or (b) confirm R04 moved those seams into the named lane. The first checkable step is a fixed two-position append/read fixture asserting FP16-round parity, resident bytes <=55% of the aligned F32 baseline, no F32 shadow cache, and `fallback_count=0`.

**Comment by anthony-chaudhary** [comment]
Provenance correction from the independent prior-art audit:

- The supplied receipt `study_fea6c1bc31f94fcee01ec566289e7ca509c28e7a982068dbde17b36e33b333e3` exists, but it is the broad Qwen3.8 comparison-ladder study and does **not** contain `lhl/strix-halo-testing` or the FP16-KV axis. It must not be treated as source-bound proof for this leaf.
- Exact issue-pinned source: `lhl/strix-halo-testing/llm-bench/README.md:10-18@5ee14b5b0ecfbd16b7fd0aa41a91581c32b33b61` (Apache-2.0). It provides only the ~215 GB/s bandwidth-budget motivation; disposition is **INSPIRE / measurement context**, not copied implementation and not speedup evidence.
- The already-studied implementation prior art is Nathanw1014's Strix Halo F16 KV contiguization (`dequant_f16_transpose.comp` plus its host gate at pinned fork commit `442d7df9d6524de20644075ebab7bb5b8eddd3a3`, MIT). Eventual R05 work should independently implement the mechanism and retain the CPU/F32 oracle, fak-native Vulkan identity, and zero-fallback gates.

No new issue or note is needed for this correction; #12388 remains the owning leaf. A source-specific immutable study receipt should be created when the dependency fence clears and the post-R04 implementation shape is known.

**Comment by anthony-chaudhary** [comment]
Bounded software-contract slice landed and published on `origin/main` via reviewed semantic reconciliation with #12399.

- Implementation: `be768e6e99cff990275d317c599251f2602864e0`; hardening: `c50e8bf3282d80047ca373ab4ef6df4904fc8167`; published merge: `533d74d0571bacaaf91d85c81ba3093fbbcfe597`.
- Module: `internal/compute@r413+gc50e8bf32`.
- Exact witness: `CGO_ENABLED=0 go test -tags vulkan ./internal/compute -run '^TestVulkanFP16KVNoF32Expansion$' -count=1 -v` — PASS.
- Contract evidence: packed binary16 K/V resident storage, no resident F32 shadow, odd-width cross-position packing, per-position ephemeral F32 widening, max F32-oracle delta `0.0009527206420898438`, aligned modeled footprint `4352` vs `8448` bytes (`48.5%` reduction), contract-local `fallback_count=0`.
- Review hardening: checked dimension/allocation arithmetic rejects adversarial overflow; invariant-bearing fields are private with read-only getters.
- Shader witness: independent WSL run used `/usr/bin/glslc` and compiled the packed specialization to a valid 5,692-byte SPIR-V module. The published semantic merge preserves both this FP16 specialization and #12399's packed-append contract plus the upstream softmax correction.
- Audits: original implementation commit is `dos commit-audit` `OK / code_effect / diff-witnessed`; two independent reviewers passed the hardening and two-parent merge.

This is intentionally **not** a runtime or physical-promotion claim: `software_contract=true`, `runtime_integrated=false`, `physical_executed=false`. Per #12492, the coordination/promotion edges did not block this bounded slice. Keep #12388 open for allocator/dispatch binding, device execution, and the exact-artifact same-envelope physical A/B promotion gate.
